lib/postgres_v3_experimental.dart

```diff
@@ -0,0 +1,258 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:collection/collection.dart';
+
+import 'src/v3/connection.dart';
+import 'src/v3/query_description.dart';
+import 'src/v3/types.dart';
+
+export 'src/v3/types.dart';
+
+abstract class PgPool implements PgSession, PgSessionExecutor {
+  static Future<PgPool> open(
+    List<PgEndpoint> endpoints, {
+    PgSessionSettings? sessionSettings,
+    PgPoolSettings? poolSettings,
+  }) =>
+      throw UnimplementedError();
+
+  Future<R> withConnection<R>(
+    Future<R> Function(PgConnection connection) fn, {
+    PgSessionSettings? sessionSettings,
+  });
+}
+
+class PgQueryDescription {
+  factory PgQueryDescription.direct(String sql, {List<PgDataType>? types}) =
+      InternalQueryDescription.direct;
+  factory PgQueryDescription.map(String sql, {String substitution}) =
+      InternalQueryDescription.map;
+}
+
+abstract class PgSession {
+  // uses extended query protocol
+  Future<PgStatement> prepare(
+    Object /* String | InternalQueryDescription */ query, {
+    Duration? timeout,
+  });
+
+  Future<PgResult> execute(
+    Object /* String | InternalQueryDescription */ query, {
+    Object? /* List<Object?|PgTypedParameter> | Map<String, Object?|PgTypedParameter> */ parameters,
+    Duration? timeout,
+  });
+
+  Future<void> close();
+}
+
+abstract class PgSessionExecutor {
+  // TODO: also add retry options similarly to postgres_pool
+  Future<R> run<R>(Future<R> Function(PgSession session) fn);
+  Future<R> runTx<R>(Future<R> Function(PgSession session) fn);
+}
+
+abstract class PgConnection implements PgSession, PgSessionExecutor {
+  static Future<PgConnection> open(
+    PgEndpoint endpoint, {
+    PgSessionSettings? sessionSettings,
+  }) {
+    return PgConnectionImplementation.connect(endpoint,
+        sessionSettings: sessionSettings);
+  }
+
+  PgChannels get channels;
+  PgMessages get messages;
+}
+
+abstract class PgResultStream implements Stream<PgResultRow> {
+  @override
+  PgResultStreamSubscription listen(void Function(PgResultRow event)? onData,
+      {Function? onError, void Function()? onDone, bool? cancelOnError});
+}
+
+abstract class PgResultStreamSubscription
+    implements StreamSubscription<PgResultRow> {
+  Future<int> get affectedRows;
+  Future<PgResultSchema> get schema;
```

Comment:
I have moved these two getters into the `StreamSubcription` interface instead of having them on `PgResultStream`.

This feels more well-behaved to me, since `Stream`s don't typically start their work until they're listened to. It also means that asynchronous errors that might occur while sending the `Bind` message can't be unhandled.